crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs

```diff
@@ -0,0 +1,882 @@
+/// always form a contiguous range with the range of the element itself;
+/// however, inline comments won't necessary form a contiguous range.
+/// Consider the following scenario, where both `# comment0` and `# comment1`
+/// will move with the "a" element when the list is sorted:
+///
+/// ```python
+/// __all__ = [
+///     "b",
+///     # comment0
+///     "a", "c",  # comment1
+/// ]
+/// ```
+///
+/// The desired outcome here is:
+///
+/// ```python
+/// __all__ = [
+///     # comment0
+///     "a",  # comment1
+///     "b",
+///     "c",
+/// ]
+/// ```
+///
+/// To achieve this, both `# comment0` and `# comment1`
+/// are grouped into the `DunderAllItem` instance
+/// where the value is `"a"`, even though the source-code range
+/// of `# comment1` does not form a contiguous range with the
+/// source-code range of `"a"`.
+#[derive(Debug)]
+struct DunderAllItem {
+    value: String,
+    preceding_comment_ranges: Vec<TextRange>,
+    element_range: TextRange,
+    // total_range incorporates the ranges of preceding comments
+    // (which must be contiguous with the element),
+    // but doesn't incorporate any trailing comments
+    total_range: TextRange,
+    end_of_line_comments: Option<TextRange>,
+}
+
+impl DunderAllItem {
+    fn new(
+        value: String,
+        preceding_comment_ranges: Vec<TextRange>,
+        element_range: TextRange,
+        end_of_line_comments: Option<TextRange>,
+    ) -> Self {
+        let total_range = {
+            if let Some(first_comment_range) = preceding_comment_ranges.first() {
+                TextRange::new(first_comment_range.start(), element_range.end())
+            } else {
+                element_range
+            }
+        };
+        Self {
+            value,
+            preceding_comment_ranges,
+            element_range,
+            total_range,
+            end_of_line_comments,
+        }
+    }
+
+    fn with_no_comments(value: String, element_range: TextRange) -> Self {
+        Self::new(value, vec![], element_range, None)
+    }
+}
+
+impl Ranged for DunderAllItem {
+    fn range(&self) -> TextRange {
+        self.total_range
+    }
+}
+
+fn multiline_dunder_all_prelude(
+    first_item_start_offset: TextSize,
+    newline: &str,
+    dunder_all_offset: TextSize,
+    locator: &Locator,
+) -> String {
+    let prelude_end = {
+        let first_item_line_offset = locator.line_start(first_item_start_offset);
+        if first_item_line_offset == locator.line_start(dunder_all_offset) {
+            first_item_start_offset
+        } else {
+            first_item_line_offset
+        }
+    };
+    let prelude = locator.slice(TextRange::new(dunder_all_offset, prelude_end));
+    format!("{}{}", prelude.trim_end(), newline)
+}
+
+fn join_multiline_dunder_all_items(
+    sorted_items: &[DunderAllItem],
+    locator: &Locator,
+    item_indent: &str,
+    newline: &str,
+    needs_trailing_comma: bool,
+) -> String {
+    let max_index = sorted_items.len() - 1;
```

Comment:
Nit: Rename to `last_item_index`. It holds slightly more information, because it gives meaning to what item the item with the `max` index points to.